### ISSUE TITLE
False warning for delegate assignments with covariant nullable conversions for ref return types

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#46977]
**Version Used**: 

**Steps to Reproduce**:
```
Branch master (18 Aug 2020)
Latest commit ca88e94 by Julien Couvreur:
Handle deconstruction variables in invalid scenarios (#46655)
```

Compile the following code
```cs
#nullable enable
delegate ref readonly T D<T>();

class C
{
    void M()
    {
        D<string> d1 = MethodGroup;
        D<string?> d2 = d1; // false warning
    }
    
    string f = "";
    ref readonly string MethodGroup() { return ref f; }
}
```

**Expected Behavior**:
No warnings

**Actual Behavior**:
```
CS8619: Nullability of reference types in value of type 'D<string>' doesn't match target type 'D<string?>'.
```
Reported for `d2 = d1` assignment because of the difference in return types

**Notes**
Roslyn allows covariant nullability conversions for ref returns in a couple of other places e.g. when overriding or for method group conversions to delegates, see https://github.com/dotnet/roslyn/issues/46971#issuecomment-676740890 for examples

Might be fixed when this PR is merged: https://github.com/dotnet/roslyn/pull/46953

**Comment by RikkiGibson** [comment]
This issue will be half-fixed by #46953. `D<string?> d2 = d1;` is an implicit reference conversion which simply looks at the variance of type parameters when deciding convertibility, not how those type parameters are used in the delegate signature, thus it will continue to warn. However something like `D<string?> d2 = MethodGroup;` or `D<string?> d2 = new D<string?>(d1)` is expected to work without warnings.
